### DISCUSSION TITLE
Creating observable that emits null is deprecated in v7?

### DISCUSSION THREAD

**Discussion by theodorejb** [ReactiveX/rxjs#5966] [Q&A]
I have a method like this in my project:

```typescript
getCurrentUserOrNull(): Observable<User | null> {
    return this.getCurrentUser()
        .pipe(
            catchError(() => of(null)),
        );
}
```

After installing RxJS 7.0.0-beta.10, there is now a deprecation notice saying "of is deprecated: The scheduler argument is deprecated, use scheduled." Is this a spurious warning since I'm not actually using a scheduler, or is the creation of observables that emit null/undefined values actually being deprecated? If the warning is spurious, is there a way to avoid triggering it for a single `null` value like this before RxJS 7 is finalized?

**Comment by kwonoj** [comment]
I don't think emitting null itself is deprecated. Not sure if we have known issues, @cartant maybe aware?

**Comment by cartant** [comment]
> I don't think emitting null itself is deprecated.

Nope, it's definitely not. But whether or not it's interpreted as a scheduler probably depends on the TypeScript settings. Are `strictNullChecks` enabled? IMO, TypeScript should never be used without that setting being enabled.

The deprecations are something that still need work before the next version is released. This is also a good example of why the deprecation messages that are used ATM are not good enough - they confuse people. IMO, we need to get the messages sorted out and have them include links to documentation pages that explain the reasoning, etc.

**Reply by kwonoj** [reply]
Sounds good and sounds like it's under umbrella issue around 7-related typing todos.

p.s unrelated note but I like new discussions having `reply` support 😍

**Reply by cartant** [reply]
Also, if `strictNullChecks` is the issue, I think we'll have to say that we don't support it being specified as `false`, as - in addition to the spurious deprecation - the inferred type will wrong, too:

https://github.com/ReactiveX/rxjs/blob/7bbd37f53397193cf1371b93c1f93b18c071474f/src/internal/observable/of.ts#L9

It would be inferred as `Observable<never>` instead of `Observable<null>`.

**Comment by benlesh** [comment]
I see. This is an issue with the type thinking `null` was a scheduler. I was going to say, the runtime behavior should not be like this at all. Maybe we should have dtslint tests for this?

**Reply by cartant** [reply]
I don't think we can have 'smart' types and also support `strictNullChecks: false`.

**Comment by Raiondesu** [comment]
I think the problem here is due to the overloads ordering for the `of` function:
https://github.com/ReactiveX/rxjs/blob/4dd7118988a17bed026b9e033947b8e57a627fc8/src/internal/observable/of.ts#L7-L70
TypeScript tries to apply overloads in order, from top to bottom.
When presented with `null` as an argument, it picks the first matching overload, which in this case seems to be the first one with the `scheduler` parameter.

This is a problem not only with `of` but with a lot of other functions that have deprecations and generics.
Many have deprecated overloads at the top, which makes TS consider them first (they should be moved to the bottom!), others do not have "general" overloads (`pipe` lacks an overload for variable amount of arguments).
